services/rpcfilters/transaction_sent_to_upstream_event.go

```diff
@@ -0,0 +1,105 @@
+package rpcfilters
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// transactionSentToUpstreamEvent represents an event that one can subscribe to
+type transactionSentToUpstreamEvent struct {
+	sxMu     sync.Mutex
+	sx       map[int]chan common.Hash
+	listener chan common.Hash
+	quit     chan struct{}
+}
+
+func newTransactionSentToUpstreamEvent() *transactionSentToUpstreamEvent {
+	return &transactionSentToUpstreamEvent{
+		sx:       make(map[int]chan common.Hash),
+		listener: make(chan common.Hash),
+	}
+}
+
+func (e *transactionSentToUpstreamEvent) Start() error {
+	if e.quit != nil {
+		return errors.New("latest transaction sent to upstream event is already started")
+	}
+
+	e.quit = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case transactionHash := <-e.listener:
+				if e.numberOfSubscriptions() == 0 {
+					continue
+				}
+				e.processTransactionSentToUpstream(transactionHash)
+			case <-e.quit:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (e *transactionSentToUpstreamEvent) numberOfSubscriptions() int {
+	e.sxMu.Lock()
+	defer e.sxMu.Unlock()
+	return len(e.sx)
+}
+
+func (e *transactionSentToUpstreamEvent) processTransactionSentToUpstream(transactionHash common.Hash) {
+
+	e.sxMu.Lock()
+	defer e.sxMu.Unlock()
+
+	for _, channel := range e.sx {
+		// send hash into channel inside goroutine so that new hashes can be sent
+		// even when not all subscribers have received the previous hash
+		ch := channel
+		go func() {
+			ch <- transactionHash
```

Comment:
Not sure if this is better than using a buffered channel. If the consumer is slow, we will keep increasing number of goroutines but we will deliver the hash to everyone. This is dangerous to the application. If we use a buffered channel, we may hit the limit of a channel but we won't increase the number of goroutines indefinitely. 

I would try to explore this solution:
```go
func (e *transactionSentToUpstreamEvent) Subscribe() (int, chan common.Hash) {
  ...
  // 512 is an arbitrary number but accumulating 512 pending txs should be enough
  channel := make(chan common.Hash, 512)
  ...
}

// change sending the hash to this
select {
case channel <- transactionHash:
default:
  log.Error("dropping messages %s for subscriotion %d because the channel is full", transactionHash, id)
}
```